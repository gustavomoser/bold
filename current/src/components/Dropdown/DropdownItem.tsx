import React from 'react'

import { focusBoxShadow, Theme, useStyles } from '../../styles'

export interface DropdownItemProps extends React.HTMLAttributes<HTMLElement> {
  type?: 'normal' | 'danger'
  disabled?: boolean
  component?: React.ElementType<any>
  [prop: string]: any
}

export function DropdownItem(props: DropdownItemProps) {
  const { type, disabled, children, onClick, component: Component, ...rest } = props
  const { classes, css } = useStyles(styles)

  const handleKeyDown = (event: React.KeyboardEvent<HTMLElement>) => {
    if (event.key === 'Enter' || event.key === ' ') {
      const elem = event.target as any
      elem.click()
    }
  }

  const handleClick = (e: React.MouseEvent<HTMLElement>) => {
    if (disabled) {
      e.stopPropagation() // avoid closing dropdown menu when disabled item is clicked
    } else {
      onClick && onClick(e)
    }
  }

  const classNames = css(classes.item, type === 'danger' && classes.danger, disabled && classes.disabled)

  return (
    <Component
      className={classNames}
      onClick={handleClick}
      onKeyDown={handleKeyDown}
      role='menuitem'
      aria-disabled={disabled ? true : undefined}
      {...rest}
    >
      {children}
    </Component>
  )
}

DropdownItem.defaultProps = {
  component: 'li',
  type: 'normal',
  disabled: false,
} as Partial<DropdownItemProps>

const styles = (theme: Theme) => ({
  item: {
    ...theme.typography.variant('main'),
    margin: 0,
    textDecoration: 'none',
    cursor: 'pointer',
    padding: 'calc(0.5rem - 2px) 1rem',
    lineHeight: '1.25rem',
    fontWeight: 'bold',
    outline: 'none',
    fontSize: theme.typography.sizes.button,
    transition: 'all .2s ease',
    display: 'block',
    '&:hover': {
      background: theme.pallete.surface.background,
    },
    '&:focus': {
      boxShadow: focusBoxShadow(theme, 'primary', 'inset'),
      borderRadius: 4,
    },
  } as React.CSSProperties,
  disabled: {
    color: theme.pallete.text.disabled,
    '&:hover': {
      cursor: 'not-allowed',
      background: 'transparent',
      color: theme.pallete.text.disabled,
    },
  },
  danger: {
    color: theme.pallete.status.danger.main,
    '&:hover': {
      color: theme.pallete.surface.main,
      background: theme.pallete.status.danger.main,
    },
    '&:focus': {
      boxShadow: focusBoxShadow(theme, 'danger', 'inset'),
      borderRadius: 4,
    },
  },
})
