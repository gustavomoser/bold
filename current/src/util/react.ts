import { MutableRefObject, Ref } from 'react'

/**
 * Compose multiple react refs.
 *
 * @returns A new ref that passes down the Element to all ref arguments.
 */
export function composeRefs<T>(...refs: Array<Ref<T>>) {
  return (value: T) => refs.forEach(ref => setRef(ref, value))
}

/**
 * Changes the value of a ref.
 * The ref can be a function ref or a object ref.
 *
 * @param ref The ref to have its value changed.
 * @param value The new value.
 */
export function setRef<T>(ref: Ref<T>, value: any) {
  if (typeof ref === 'function') {
    ref(value)
  } else if (ref) {
    const mutableRef = ref as MutableRefObject<T>
    mutableRef.current = value
  }
}

/**
 * Compose multiple event handlers.
 *
 * @returns A new function that triggers all handlers passed as parameters, in order.
 */
export function composeHandlers(...handlers: Array<(...params: any) => void>) {
  return (...args: any[]) => handlers.forEach(handler => handler && handler(...args))
}

/**
 * Invokes the native `value` property setter of an element.
 * From https://github.com/facebook/react/issues/10135#issuecomment-401496776
 *
 * @param element The element to invoke the setter on.
 * @param value The value to be set.
 */
export function setNativeValue(element: HTMLElement, value: any) {
  const prototype = Object.getPrototypeOf(element)
  const { set: prototypeValueSetter = null } = Object.getOwnPropertyDescriptor(prototype, 'value') || {}
  const { set: valueSetter = null } = Object.getOwnPropertyDescriptor(element, 'value') || {}

  if (prototypeValueSetter && valueSetter !== prototypeValueSetter) {
    prototypeValueSetter.call(element, value)
  } else if (valueSetter) {
    valueSetter.call(element, value)
  } else {
    throw new Error('The given element does not have a value setter')
  }
}
