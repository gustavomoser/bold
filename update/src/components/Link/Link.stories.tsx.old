import { text } from '@storybook/addon-knobs'
import React from 'react'

import { Link } from './Link'

export default {
  title: 'Components/Link',
}

export const Default = () => (
  <Link href={text('href', '/')} target='_blank'>
    {text('children', 'Link to somewhere')}
  </Link>
)
