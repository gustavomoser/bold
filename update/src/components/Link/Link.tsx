import React, { AnchorHTMLAttributes, CSSProperties } from 'react'

import { focusBoxShadow, Theme, useStyles } from '../../styles'
import { Omit } from '../../util'
import { BaseTextProps, Text } from '../Text'

export type LinkProps = Omit<AnchorHTMLAttributes<HTMLAnchorElement>, 'style'> & BaseTextProps

export function Link<P = LinkProps>(props: P & BaseTextProps) {
  const { classes } = useStyles(createStyles, props)

  return <Text<LinkProps> variant='link' component='a' {...props} style={[classes.link, props.style]} />
}

export const createStyles = (theme: Theme) => ({
  link: {
    cursor: 'pointer',
    outline: 'none',
    transition: 'box-shadow .2s ease',
    '&:hover': {
      textDecoration: 'none',
    },
    '&:focus': {
      textDecoration: 'none',
      borderRadius: theme.radius.input,
      boxShadow: focusBoxShadow(theme),
    },
  } as CSSProperties,
})
