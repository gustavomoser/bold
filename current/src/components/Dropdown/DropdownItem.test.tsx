import { fireEvent, render } from '@testing-library/react'
import React from 'react'

import { Link } from '../Link'

import { DropdownItem } from './DropdownItem'

it('should render correctly', () => {
  const { container } = render(<DropdownItem>Item</DropdownItem>)
  expect(container).toMatchSnapshot()
})

it('should accept "component" prop to render with different container elements', () => {
  const { container: container1 } = render(
    <DropdownItem component='a' href='/'>
      Item
    </DropdownItem>
  )
  expect(container1).toMatchSnapshot()

  const { container: container2 } = render(
    <DropdownItem component={Link} href='/'>
      Item
    </DropdownItem>
  )
  expect(container2).toMatchSnapshot()
})

describe('onClick', () => {
  it('should call onClick when clicked', () => {
    const handleClick = jest.fn()
    const { container } = render(<DropdownItem onClick={handleClick}>Item</DropdownItem>)
    fireEvent.click(container.querySelector('li'))
    expect(handleClick).toHaveBeenCalled()
  })

  it('should NOT call onClick when disabled', () => {
    const handleClick = jest.fn()
    const { container } = render(
      <DropdownItem onClick={handleClick} disabled>
        Item
      </DropdownItem>
    )
    fireEvent.click(container.querySelector('li'))
    expect(handleClick).not.toHaveBeenCalled()
  })

  it('should call onClick when focused and Enter is pressed', () => {
    const handleClick = jest.fn()
    const { container } = render(<DropdownItem onClick={handleClick}>Item</DropdownItem>)
    fireEvent.focus(container.querySelector('li'))
    fireEvent.keyDown(container.querySelector('li'), { key: 'Enter' })
    expect(handleClick).toHaveBeenCalled()
  })

  it('should call onClick when focused and Spacebar is pressed', () => {
    const handleClick = jest.fn()
    const { container } = render(<DropdownItem onClick={handleClick}>Item</DropdownItem>)
    fireEvent.focus(container.querySelector('li'))
    fireEvent.keyDown(container.querySelector('li'), { key: ' ' })
    expect(handleClick).toHaveBeenCalled()
  })
})
