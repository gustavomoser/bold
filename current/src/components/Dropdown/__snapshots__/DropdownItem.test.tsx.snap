// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept "component" prop to render with different container elements 1`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.875rem;
  line-height: 1.25rem;
  color: #24252E;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  padding: calc(0.5rem - 2px) 1rem;
  font-weight: bold;
  outline: none;
  -webkit-transition: all .2s ease;
  transition: all .2s ease;
  display: block;
}

.emotion-0:hover {
  background: #F0F0F5;
}

.emotion-0:focus {
  box-shadow: inset 0 0 0 2px #0069D0;
  border-radius: 4px;
}

<div>
  <a
    class="emotion-0"
    href="/"
    role="menuitem"
  >
    Item
  </a>
</div>
`;

exports[`should accept "component" prop to render with different container elements 2`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.875rem;
  line-height: 1.25rem;
  color: #24252E;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  padding: calc(0.5rem - 2px) 1rem;
  font-weight: bold;
  outline: none;
  -webkit-transition: all .2s ease;
  transition: all .2s ease;
  display: block;
}

.emotion-0:hover {
  background: #F0F0F5;
}

.emotion-0:focus {
  box-shadow: inset 0 0 0 2px #0069D0;
  border-radius: 4px;
}

<div>
  <a
    class="emotion-0"
    href="/"
    role="menuitem"
  >
    Item
  </a>
</div>
`;

exports[`should render correctly 1`] = `
.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.875rem;
  line-height: 1.25rem;
  color: #24252E;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  padding: calc(0.5rem - 2px) 1rem;
  font-weight: bold;
  outline: none;
  -webkit-transition: all .2s ease;
  transition: all .2s ease;
  display: block;
}

.emotion-0:hover {
  background: #F0F0F5;
}

.emotion-0:focus {
  box-shadow: inset 0 0 0 2px #0069D0;
  border-radius: 4px;
}

<div>
  <li
    class="emotion-0"
    role="menuitem"
  >
    Item
  </li>
</div>
`;
